### DISCUSSION TITLE
How to Manage Multiple Database Sessions when working with an API including Async Tasks

### DISCUSSION THREAD

**Discussion by gauravdh9** [fastapi/fastapi#8347] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class RDSConnector(metaclass=Singleton):

    Base = declarative_base()

    async def intialize_rds_connection(self, db_config):
        # read connection
        read_pool_size_count = db_config.get("read_rds").get("pool_size")
        read_max_overflow_count = db_config.get("read_rds").get("max_overflow")
        db_url = 'mysql+pymysql'+'://'+read_user+':'+read_pwd+'@'+host+':'+str(port)+'/'+db_name
        engine = create_engine(db_url, pool_recycle=10, pool_pre_ping=True, pool_size=read_pool_size_count, max_overflow=read_max_overflow_count,pool_timeout=20)
        self.sessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

        # write connection
        write_pool_size_count = db_config.get("write_rds").get("pool_size")
        write_max_overflow_count = db_config.get("write_rds").get("max_overflow")
        wdb_url = 'mysql+pymysql'+'://'+write_user+':'+write_pwd+'@'+whost+':'+str(wport)+'/'+wdb_name
        write_engine = create_engine(wdb_url, pool_recycle=10, pool_pre_ping=True, pool_size=write_pool_size_count, max_overflow=write_max_overflow_count,pool_timeout=20)
        self.writeSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=write_engine)



    def get_db_read(self):
        db = self.sessionLocal()
        try:
            #logger.info('db connected')
            yield db
        finally:
            db.close()
            #logger.info('db close')
    
    def get_db_write(self):
        db = self.writeSessionLocal()
        try:
            #logger.info('db connected')
            yield db
        finally:
            db.close()
            #logger.info('db close')
```

### Description

By default When an EndPoint receives a request with help of Dependency Injection I create a DB session for that route and use it inside that Service/Route/DAO Layer.
Consider an API/endpoint having multiple reads and writes. 
Now because pool_recycle is set to 20 and RDS connection in production environment is set to 1 minute after one minute connection would be lost.
If an Asynchronous task takes 30 Minutes to complete and on that completion I need to make some write. I wouldn’t  be able to make because i don’t have active db session.
### Also I need to make sure DB session has both read and write db's Instance

# Probable Way of direction of Solution:-
1. Context Managers for DB Sessions:
2. DBSessionMiddleWare
3. Routing Class to manage both read and write

### Operating System

macOS

### FastAPI Version

0.75.0

### Python Version

3.8.9

**Comment by jgould22** [comment]
> Now because pool_recycle is set to 20 and RDS connection in production environment is set to 1 minute after one minute connection would be lost.

Have you tested that this is correct? I do not believe that it is.

Connections are only invalidated during checkout. So long as you have the connection checked out it will not be recycled.  When you return it to the the pool the it will be recycled.

https://docs.sqlalchemy.org/en/14/core/pooling.html#setting-pool-recycle
